### ISSUE TITLE
Type narrowing with `type(...)` does not work in all cases

### ISSUE THREAD

**Issue by shiftinv** [microsoft/pyright#3681]
**Describe the bug**
When using `isinstance` with a `type()` as the second argument, the variable type is correctly narrowed in the positive case but not in the negative case.

**To Reproduce**
```py
from __future__ import annotations

import sys
from typing import Optional, Union

if sys.version_info >= (3, 10):
    from types import EllipsisType
else:
    EllipsisType = type(Ellipsis)


def func(param: Union[int, EllipsisType]) -> Optional[int]:
    reveal_type(param)
    if isinstance(param, EllipsisType):
        reveal_type(param)
        return None

    reveal_type(param)
    return param
```

```sh
$ pyright --pythonversion 3.9 test.py  # uses type(Ellipsis) fallback
...
test.py
  test.py:13:17 - information: Type of "param" is "int | ellipsis"
  test.py:15:21 - information: Type of "param" is "ellipsis"
  test.py:18:17 - information: Type of "param" is "int | ellipsis"
  test.py:19:12 - error: Expression of type "int | EllipsisType" cannot be assigned to return type "int | None"
    Type "int | EllipsisType" cannot be assigned to type "int | None"
      Type "EllipsisType" cannot be assigned to type "int | None"
        "EllipsisType" is incompatible with "int"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
1 error, 0 warnings, 3 informations

$ pyright --pythonversion 3.10 test.py  # uses types.EllipsisType
...
test.py
  test.py:13:17 - information: Type of "param" is "int | ellipsis"
  test.py:15:21 - information: Type of "param" is "ellipsis"
  test.py:18:17 - information: Type of "param" is "int"
0 errors, 0 warnings, 3 informations
```

**Expected behavior**
`type(Ellipsis)` should work the same way as `types.EllipsisType` for type-checking.

**Screenshots or Code**
-

**VS Code extension or command-line**
cli, 1.1.258

**Additional context**
This does not seem specific to `EllipsisType` and can also be reproduced with `str` and `type("")`, though there's no point in using `type("")` over `str`.

**Comment by erictraut** [comment]
This behavior is intentional. In one case `EllipsisType` refers to a class, `ellipsis`. In the other, `EllipsisType` comes from passing `Ellipsis` through the `type` constructor. `Ellipsis` is a variable whose type is declared as `ellipsis`. That means its actual type could be a subtype thereof. (In this particular example, `ellipsis` happens to be marked `@final`, but in the general case, `x: Y` means that variable `x` is of type `Y` or a subtype thereof.) When pyright performs isinstance type narrowing, it differentiates between these two cases. If it knows that a type refers specifically to a class (like `ellipsis`) 

Consider the following example:

```python
class Parent: ...
class Child(Parent): ...

def func1(a: Parent | int):
    if isinstance(a, Parent):
        reveal_type(a) # Parent
    else:
        reveal_type(a) # int

def func2(a: Parent | int, b: type[Parent]):
    if isinstance(a, b):
        reveal_type(a) # Parent
    else:
        reveal_type(a) # Parent | int
```

In the case of `func1`, the second argument to `isinstance` refers to class `Parent`. In the case of `func2`, the second argument to `isinstance` refers to some type that is `Parent` or a subtype thereof. The variable `b` could be `Parent`, but it could also be `Child`. In the latter case, it is not safe to narrow `a` in the negative (else) case.

Mypy reveals the same types in the example above.

Your code sample is the same thing except with `ellipsis`.

So this isn't a bug. Pyright is working as expected in this case.

**Comment by shiftinv** [comment]
Thanks for the explanation! That does make sense - I figured this could be intended behavior since, as you mentioned, the type of `Ellipsis` may technically be a subtype of `ellipsis`, but I wasn't sure if that might be an edge case that's worth handling since `ellipsis` is marked as `@final`, and does not exist at runtime at all.
In that case, I suppose there's no alternative that does not involve `ellipsis`. It's a bit unfortunate as tools like flake8 naturally don't like the use of `ellipsis`; even typeshed has a `# noqa: F821` on the `EllipsisType = ellipsis` definition, for our purposes a `noqa` will do just fine here then 😅
